File: solutions/javascript/word-ladder.js
function ladderLength(begin, end, wordList) {
  const wordSet = new Set(wordList);
  if (!wordSet.has(end)) return 0;

  const queue = [[begin, 1]];
  while (queue.length) {
    const [word, length] = queue.shift();
    if (word === end) return length;

    for (let i = 0; i < word.length; i++) {
      for (let c of "abcdefghijklmnopqrstuvwxyz") {
        const newWord = word.slice(0, i) + c + word.slice(i+1);
        if (wordSet.has(newWord)) {
          wordSet.delete(newWord);
          queue.push([newWord, length + 1]);
        }
      }
    }
  }
  return 0;
}
console.log(ladderLength("hit","cog",["hot","dot","dog","lot","log","cog"]));

File: solutions/java/WordLadder.java
import java.util.*;

public class WordLadder {
    public static int ladderLength(String begin, String end, java.util.List<String> wordList) {
        Set<String> wordSet = new HashSet<>(wordList);
        if (!wordSet.contains(end)) return 0;

        Queue<String> q = new LinkedList<>();
        q.offer(begin);
        int level = 1;

        while (!q.isEmpty()) {
            int size = q.size();
            for (int i=0;i<size;i++) {
                String word = q.poll();
                if (word.equals(end)) return level;

                char[] arr = word.toCharArray();
                for (int j=0;j<arr.length;j++) {
                    char original = arr[j];
                    for (char c='a';c<='z';c++) {
                        arr[j]=c;
                        String newWord = new String(arr);
                        if (wordSet.contains(newWord)) {
                            wordSet.remove(newWord);
                            q.offer(newWord);
                        }
                    }
                    arr[j]=original;
                }
            }
            level++;
        }
        return 0;
    }

    public static void main(String[] args) {
        System.out.println(ladderLength("hit","cog",java.util.Arrays.asList("hot","dot","dog","lot","log","cog")));
    }
}